config/initializers/validators/certificate_validator.rb

```diff
@@ -23,19 +23,31 @@ def certificate_is_valid
     certificate_key_is_supported
   end
 
-  def factory
-    Utilities::Certificate::CertificateFactory
+  def convert_value_to_x509_certificate
+    begin
+      OpenSSL::X509::Certificate.new(value)
+    rescue
+      OpenSSL::X509::Certificate.new(Base64.decode64(value))
+    end
   end
 
   def x509_certificate
-    x509 = factory.x509_certificate(value)
+    if value != @last_converted_value || @x509.blank?
```

Comment:
Code was originally written by Kerr but I and Rosa wondered what it was actually doing and came up with:
@last_converted_value performs some short circuiting, for instance lets say i put a valid cert value during upload but didn't click on signing or encryption radio item `@last_converted = value` . it's not going to perform the upload because one of the options have to have been selected. 
The second time around I pick on option, since the value is stored in @last_converted this whole section is skipped and provide the cert value is fine the upload takes place
``` if value != @last_converted_value || @x509.blank?
        @x509 = convert_value_to_x509_certificate
        @last_converted_value = value
    end``
and the stored value in @x509 is used.
But say during my second go; the value i generated is not valid and I know this because maybe the end date is used to generate the value is over say a year. So I addition to selecting an option (encryption or signing) I also go ahead and regenerate a value which a paste over the the existing one then value != @last_converted_value and the `convert_value_to_x509_certificate` run ensuring the my regenerated value is picked up.

I don't know if there is better why to do this sort of lazy evaluation

Reply:
>  put a valid cert value during upload but didn't click on signing or encryption radio item
in that case we should change the order of the validation, if the form is not valid, we shouldn't even attempt to validate the actual data/cert

> The second time around I pick on option, since the value is stored in @last_converted
is it really stored? where would it be, the app is stateless and if we have multiple instances running it won't be there.